FWCore/Framework/interface/SignallingProductRegistryFiller.h

```diff
@@ -0,0 +1,114 @@
+#ifndef FWCore_Framework_SignallingProductRegistryFiller_h
+#define FWCore_Framework_SignallingProductRegistryFiller_h
+// -*- C++ -*-
+//
+// Package:     Framework
+// Class  :     SignallingProductRegistryFiller
+//
+/**\class SignallingProductRegistryFiller SignallingProductRegistryFiller.h FWCore/Framework/interface/SignallingProductRegistryFiller.h
+
+ Description: <one line class summary>
+
+ Usage:
+    <usage>
+
+*/
+//
+// Original Author:  Chris Jones
+//         Created:  Fri Sep 23 16:47:10 CEST 2005
+//
+
+// system include files
+#include <map>
+#include <string>
+
+#include "FWCore/Utilities/interface/Signal.h"
+
+// user include files
+#include "DataFormats/Provenance/interface/ProductRegistry.h"
+#include "FWCore/ServiceRegistry/interface/connect_but_block_self.h"
+
+// forward declarations
+namespace edm {
+  class SignallingProductRegistryFiller {
+  public:
+    SignallingProductRegistryFiller() : productAddedSignal_(), registry_(), typeAddedStack_() {}
+
+    struct FromEarlierProcess {};
+    explicit SignallingProductRegistryFiller(ProductRegistry::ProductList const& preg, FromEarlierProcess)
+        : productAddedSignal_(), registry_(preg, false), typeAddedStack_() {}
+
+    signalslot::Signal<void(ProductDescription const&)> productAddedSignal_;
+
+    explicit SignallingProductRegistryFiller(ProductRegistry const& preg)
+        : productAddedSignal_(), registry_(preg), typeAddedStack_() {};
+
+    void addProduct(ProductDescription const& productdesc, bool iFromListener = false) {
+      registry_.addProduct_(productdesc);
+      addCalled(productdesc, iFromListener);
+    }
+
+    void addLabelAlias(ProductDescription const& productdesc,
+                       std::string const& labelAlias,
+                       std::string const& instanceAlias) {
+      addCalled(registry_.addLabelAlias_(productdesc, labelAlias, instanceAlias), false);
+    }
+    void addFromInput(edm::ProductRegistry const& iReg) {
+      registry_.addFromInput_(iReg, [this](auto const& prod) { addCalled(prod, false); });
+    }
+
+    //NOTE: this is not const since we only want items that have non-const access to this class to be
+    // able to call this internal iteration
+    // Called only for branches that are present (part of avoiding creating type information for dropped branches)
+    template <typename T>
+    void callForEachBranch(T const& iFunc) {
+      //NOTE: If implementation changes from a map, need to check that iterators are still valid
+      // after an insert with the new container, else need to copy the container and iterate over the copy
+      for (ProductRegistry::ProductList::const_iterator itEntry = registry_.productList().begin(),
+                                                        itEntryEnd = registry_.productList().end();
+           itEntry != itEntryEnd;
+           ++itEntry) {
+        if (itEntry->second.present()) {
+          iFunc(itEntry->second);
+        }
+      }
```

Comment:
Could become
```suggestion
      for (auto const& entry : registry_.productList()) {
        if (entry.second.present()) {
          iFunc(entry.second);
        }
      }
```

Reply:
I was originally concerned that the _range based for loop_ might not have the same semantics as the iterator based one and I needed that semantics since the container might be updating while the loop is happening. A bit of research and I believe the semantics are retained so I changed the code.